code/globalincs/pool.h

```diff
@@ -0,0 +1,374 @@
+	 * @return pool_index
+	 */
+	pool_index use_free(T &in){
+		auto i = known_empty.back();
+		Assertion(storage[i].stored == tl::nullopt ,"Attempted to use_free on a slot that was not free");
+		//values are created as nullopt when expanding a list
+		//and resest to nullopt here so we need to construct a new object to go in there.
+		storage[i].stored = tl::optional<T>(in);
+		pool_index r((int) i,(int) storage[i].generation);
+		known_empty.pop_back();
+		return r;
+	}
+
+	//Internal function to fill out the empty list when the storage is known to be clear.
+	void fill_empty_list(){
+		known_empty.clear();
+		if((int) storage.size()>0){
+			for (int i=((int) storage.size())-1; i>=0;i--){
+				known_empty.push_back(i);
+			}
+		}
+	}
+	//Internal to reset a slot, assumes generation checking done by interface;
+	
+	void reset_entry(gEntry &e)
+	{
+		 
+		if(e.stored == tl::nullopt)
+			return;
+		e.stored.reset();
+		e.generation++;
+	}
+	void reset_slot(int i)
+	{
+		gEntry &e = storage[i];
+		reset_entry(e);
+		known_empty.push_back(i);
+	}
+	public:
+	SCP_Pool<T>() = default;
+	SCP_Pool<T>(size_t cap_in){
+		reset(cap_in);
+	};
+
+	void reset(){
+		capped = false;
+		cap = 0;
+		for(gEntry &entry: storage)
+		{
+			reset_entry(entry);
+		}
+		fill_empty_list();
+	}
+
+	void reset(size_t cap_in){
+		capped = true;
+		cap=MAX(cap_in,storage.size());//shrinking a container is illegal.
+
+		for(gEntry entry: storage){
+			if (entry.stored != tl::nullopt)
+				entry.stored.reset();
+		}
+		for (size_t i=storage.size(); i<cap;i++){
+			gEntry e;
+			e.generation = 0;
+			e.stored = tl::nullopt;
+			storage.push_back(e);
+		}
+		fill_empty_list();
+	}
+
+	//Index into the storage... bare reference
+	T& operator[](pool_index i) {
+		if (i.null())
+			throw PoolExceptionNullIndex("null index SCP_Pool access attempted");
+		if (i.i() >= storage.size())
+			throw PoolExceptionOutOfRange("out of range SCP_Pool access attempted");
+		if (storage[i.i()].generation!= i.g())
+			throw PoolExceptionStaleIndex("Stale SCP_Pool access attempted");
+		if(storage[i.i()].stored==tl::nullopt)
+			throw PoolExceptionNullValue("Empty SCP_Pool access attempted");
+//		v = (storage[i.i()].stored).operator->();
+		return storage[i.i()].stored.operator*();
+	};
+	bool value_at(pool_index i){
+		return !(i.null()
+			|| i.i() >= storage.size()
+			|| storage[i.i()].generation!= i.g()
+			|| storage[i.i()].stored==tl::nullopt);
+	}
+	tl::optional<T&> get(pool_index i){
+		SCP_Pool<T> &self = this;
+
+		if(! value_at(i))
+			return tl::nullopt;
+		T& r = self[i];
+		return tl::optional<T&>(r);
+
+	};
+	tl::optional<T*> get_pointer(pool_index i){
+		if(! value_at(i))
+			return tl::nullopt;
+		tl::optional<T> s = storage[i.i()].stored;
+		T* p = s.operator->();
+		return p;
+	};
+
+	tl::optional<pool_index> get_new(){
+		T n = T();
+		return store(n);
+	};
+
+	tl::optional<pool_index> get_new(T&& args){
+		T n = T(std::forward<T>(args));
+		return store(n);
+	};
```

Comment:
Close, but now you still expect someone to construct the element for you (and forwarding that), instead of forwarding the args themselves to the constructor.
Effectively, I'd expect something like:
```cpp
template<typename... args_t>
tl::optional<pool_index> get_new(args_t&&... args){
	return store(T(std::forward<args_t>(args)...));
};
```